R/predict.R

```diff
@@ -0,0 +1,512 @@
+#' @noRd
+create_newdata_dataframe <- function(
+    object,
+    parameter,
+    min_date,
+    max_date,
+    horizon,
+    mean_delay,
+    gi_pmf,
+    delta = delta,
+    call = rlang::caller_env()) {
+  desired_dates <- parse_predict_dates(
+    object = object,
+    min_date = min_date,
+    max_date = max_date,
+    horizon = horizon,
+    call = call
+  )
+  dates <- shift_desired_dates(
+    parameter,
+    min(desired_dates),
+    max(desired_dates),
+    mean_delay = mean_delay,
+    gi_pmf = gi_pmf
+  )
+  timesteps <- dates_to_timesteps(
+    dates,
+    min_supplied_date = object[["min_date"]],
+    max_supplied_date = object[["max_date"]]
+  )
+  if (parameter == "r") {
+    timesteps <- interleave(timesteps - delta, timesteps + delta)
+    desired_dates <- interleave(desired_dates, NA)
+  } else if (parameter == "Rt") {
+    # Use NA padding to mark incomplete sections to remove
+    na_pad <- rep(NA, length(gi_pmf))
+    desired_dates <- as.Date(
+      c(na_pad, desired_dates, na_pad)
+    )
+  }
+  format_newdata_dataframe(
+    fit = object,
+    parameter = parameter,
+    desired_dates = desired_dates,
+    timesteps = timesteps
+  )
+}
+
+#' Format predicted data frame with uniform minimal specification
+#' @noRd
+format_predicted_dataframe <- function(
+    parameter,
+    fitted,
+    newdata,
+    delta,
+    gi_pmf) {
+  if (parameter == "r") {
+    # Difference calculation for `r` parameter
+    timestep_first_row <- which((fitted[[".row"]] - 1) %% 2 == 0)
+    fitted <- data.frame(
+      .row = (fitted[timestep_first_row, ".row"] + 1) / 2,
+      .response = discrete_diff_derivative(fitted[[".fitted"]]),
+      .draw = fitted[timestep_first_row, ".draw"]
+    )
+  } else if (parameter == "Rt") {
+    # Rt calculation
+    fitted <- rt_by_group(fitted,
+      group_cols = ".draw",
+      value_col = ".fitted",
+      vec = gi_pmf
+    )
+  }
+
+  # Merge with newdata and select required columns
+  merged <- merge(fitted,
+    newdata,
+    by = ".row"
+  )[, c(
+    "reference_date",
+    ".response",
+    ".draw"
+  )]
+  merged$reference_date <- as.Date(merged$reference_date)
+  merged[which(!is.na(merged[["reference_date"]])), ]
+}
+
+#' Turn input parameters into a dataframe to pass to model
+#' @noRd
+format_newdata_dataframe <- function(
+    fit,
+    parameter,
+    desired_dates,
+    timesteps,
+    gi_pmf,
+    delta) {
+  newdata <- gratia::data_slice(fit[["model"]], timestep = timesteps)
+
+  if (parameter == "r") {
+    # Join key when adding metadata to posterior sim. Setting
+    # every other row to NA ensures we get one row per date-draw.
+    newdata[".row"] <- interleave(seq(1, length(timesteps) / 2, 1), NA)
+  } else {
+    newdata[".row"] <- seq_along(timesteps)
+  }
+  newdata["reference_date"] <- desired_dates
+  return(newdata)
+}
+
+#' Centered difference derivative.
+#' @noRd
+discrete_diff_derivative <- function(vals) {
+  len <- length(vals)
+  t0 <- seq(1, len - 1, 2)
```

Comment:
Flagging that this might be wrong. It's meant to be centered and over a length window of two days, so it should maybe be [0, 2], [1, 3], [2, 4], ...

Reply:
But it matches the setup on 480, so there's world in which past me is right and I just need to write better docs on "centered"

Reply:
Ok past me was right. Cleaned it up here: https://github.com/CDCgov/cfa-gam-rt/pull/48/commits/940ca78b7d77cb3e31e118a18dd7f3fa8c9a75f1